applications: nrf_desktop: Add basic HID SCI support on dongle (#30315)
nrfconnect/sdk-nrf: applications/nrf_desktop/doc/ble_conn_params.rst

The connection parameter change is reverted when USB is active or disconnected.
The :option:`CONFIG_DESKTOP_BLE_USB_MANAGED_CI` is enabled by default.

> Enable :option:`CONFIG_DESKTOP_BLE_DONGLE_HID_SCI_ENABLE` to use HID SCI (Shorter Connection Intervals) on the dongle.

Reviewer: Shouldn't we move this to HID forward doc (as it's now HID forward's Kconfig option)? We could still link to the BLE conn params module docs from there

Author: done

> The option selects the HOGP SCI client support and configures default connection rate parameters used for new Bluetooth connections.

Reviewer: Consider adding `HOGP SCI client support` Kconfig option reference

Author: done

> The option selects HOGP SCI client support (:option:`CONFIG_BT_HOGP_SCI`) and configures default connection rate parameters used for new Bluetooth connections.

Reviewer 2: ```suggestion
The option selects HOGP SCI client support (:kconfig:option:`CONFIG_BT_HOGP_SCI`) and configures default connection rate parameters used for new Bluetooth connections.
```
fixes the doc build failure.

> See the "HID SCI" section in the implementation details for the related application events.

Reviewer 3: ```suggestion
See the :ref:`nrf_desktop_ble_conn_params_hid_sci' section in the implementation details for the related application events.
```

> See the :ref:`nrf_desktop_ble_conn_params_hid_sci' section in the implementation details for the related application events.

Reviewer 2: ```suggestion
See the :ref:`nrf_desktop_ble_conn_params_hid_sci` section in the implementation details for the related application events.
```

Author: done

.. note::
   On the peripheral side, the Bluetooth connection latency is controlled by :ref:`nrf_desktop_ble_latency`.

.. _nrf_desktop_ble_conn_params_hid_sci:

HID SCI
=======

> When the :option:`CONFIG_DESKTOP_HID_FORWARD_HID_SCI_ENABLE` Kconfig option is enabled, the |ble_conn_params| sets default connection rate parameters on module initialization using :c:func:`bt_conn_le_conn_rate_set_defaults`.

Reviewer: ```suggestion
When the :option:`CONFIG_DESKTOP_HID_FORWARD_HID_SCI_ENABLE` Kconfig option is enabled, the |ble_conn_params| sets promptless :option:`CONFIG_DESKTOP_BLE_CONN_PARAMS_SCI_ENABLE` Kconfig option.
The module sets default connection rate parameters on module initialization using :c:func:`bt_conn_le_conn_rate_set_defaults`.
```

.. note::
   On the peripheral side, the Bluetooth connection latency is controlled by :ref:`nrf_desktop_ble_latency`.

.. _nrf_desktop_ble_conn_params_hid_sci:

HID SCI
=======

> When the :option:`CONFIG_DESKTOP_HID_FORWARD_HID_SCI_ENABLE` Kconfig option is enabled, the |ble_conn_params| sets the promptless :option:`CONFIG_DESKTOP_BLE_CONN_PARAMS_SCI_ENABLE` Kconfig option.

Reviewer: `CONFIG_DESKTOP_BLE_CONN_PARAMS_HID_SCI_ENABLE`?

Author: done

.. note::
   On the peripheral side, the Bluetooth connection latency is controlled by :ref:`nrf_desktop_ble_latency`.

> HID SCI

Reviewer 3: ```suggestion
.. _nrf_desktop_ble_conn_params_hid_sci:

HID SCI
```

=======

When the :option:`CONFIG_DESKTOP_BLE_DONGLE_HID_SCI` Kconfig option is enabled, the |ble_conn_params| sets default connection rate parameters on module initialization using :c:func:`bt_conn_le_conn_rate_set_defaults`.

For HID SCI connections, the module does not perform the standard connection parameter update described in :ref:`nrf_desktop_ble_conn_params_connection_interval_update`.

> Instead, it controls the connection parameters by requesting appropriate HID SCI modes from the peripheral.

Reviewer: Consider mentioning that it's required by the HID SCI profile spec

Author: done

> If the peripheral supports both HID SCI and LLPM, HID SCI takes precedence.

Reviewer: Wondering if we should mention that here (we discussed that peripheral would support either SCI or LLPM but not both)

Author: Ok, removed for now

But I will still leave the part responsible for this in the source code, as we've also discussed that if we want to support the LLPM to SCI update in the future the peripheral would have to support both LLPM and SCI at the same time (as the dongle is updated before the peripheral);

Reviewer: Agreed. I am just not sure if we can assume update order (e.g. dongle can be plugged and no peripherals connected during DFU or a peripheral might be connected directly over USB).

Reviewer: Also we might consider temporarily using "regular" BLE as a fallback (making sure at some point we will switch to SCI)

> Instead, it controls the connection parameters by requesting appropriate HID SCI modes from the peripheral, as required by the `HID Over GATT Profile Specification`_.

Reviewer: ```suggestion
For HID SCI connections, the module does not perform the standard connection parameter update.
Instead, it controls the connection parameters by requesting appropriate HID SCI modes from the peripheral, as required by the `HID Over GATT Profile Specification`_.
For details, see the :ref:`nrf_desktop_ble_conn_params_connection_interval_update` section.
```

Reviewer: This use-case is covered by the linked section too.

The LLPM connection parameters update requires using vendor-specific HCI commands.
Moreover, the peripheral cannot request the LLPM connection parameters using Zephyr Bluetooth® API.

.. _nrf_desktop_ble_conn_params_connection_interval_update:

Connection interval update
==========================

After the :ref:`nrf_desktop_ble_discovery` completes the peripheral discovery, the |ble_conn_params| updates the connection parameters in the following manner:

> * If the peripheral supports HID SCI, the parameter update is skipped and the HID SCI mode is set

Reviewer: `... and HID SCI FAST mode is requested as soon a the peripheral discovery completes.`?

Author: done
